### PR TITLE
Project submitted

### PR DESCRIPTION
**Pull request by bhollan**

### CODE REVIEW

File: bin/tictactoe [hunk 1]
```diff
@@ -1,3 +1,35 @@
...
+      g.exit
+    end    
+  elsif input == "0"
+    g = Game.new(Computer.new("X", true), Computer.new("O", true))
+    g.play
+    g.exit
+  else
+    puts "Try choosing from the options next time."
+    Game.new.exit
+  end
```
Comment (randallreedjr) [inline bin/tictactoe]: Prefer to encapsulate CLI logic into a class. For example, the only code in your bin executable should be:

```ruby
#!/usr/bin/env ruby
       
require_relative '../config/environment'

TicTacToeCLI.new.call
```

And then you'd have a file: `lib/tic_tac_toe_cli.rb` with:

```ruby
class TicTacToeCLI
  def initialize
  end

  def call
   # Your game logic
  end
end
```

The code in your bin would generally just go in `call`.

File: bin/tictactoe [hunk 2]
```diff
@@ -1,3 +1,35 @@
...
+      g = Game.new(Human.new("X"), Computer.new("O", true))
+      g.play
+      g.exit
+    end    
+  elsif input == "0"
+    g = Game.new(Computer.new("X", true), Computer.new("O", true))
+    g.play
+    g.exit
+  else
+    puts "Try choosing from the options next time."
```
Comment (randallreedjr) [inline bin/tictactoe]: haha

File: lib/board.rb [hunk 3]
```diff
@@ -0,0 +1,52 @@
+class Board
+  attr_accessor :cells
+
+  def initialize
+    @cells = [" ", " ", " ", " ", " ", " ", " ", " ", " "]
+  end
+
+  def reset!
+    self.cells = [" ", " ", " ", " ", " ", " ", " ", " ", " "]
```
Comment (randallreedjr) [inline lib/board.rb]: Another way to instantiate an Array with `i` values is `Array.new(9, " ")` (see [Array.new](http://ruby-doc.org/core-2.2.0/Array.html#method-c-new)), where `9` is the number of elements in the array and `" "` is the default value for the elements. It's more concise and will prevent errors from repeatedly typing `" "`.

File: lib/board.rb [hunk 4]
```diff
@@ -0,0 +1,52 @@
+class Board
+  attr_accessor :cells
+
+  def initialize
+    @cells = [" ", " ", " ", " ", " ", " ", " ", " ", " "]
```
Comment (randallreedjr) [inline lib/board.rb]: I would just call `self.reset!` here and delegate to that method to create the default empty board.

File: lib/board.rb [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    puts "-----------"
+    puts " #{@cells[6]} | #{@cells[7]} | #{@cells[8]} "
+  end
+
+  def position(num)
+    "#{cells[num.to_i-1]}"
+  end
+
+  def full?
+    self.cells.all? {|cell| cell == "X" || cell == "O"} ? true : false
```
Comment (randallreedjr) [inline lib/board.rb]: Avoid using the ternary operator to return `true` and `false`. Instead, rely on the implicit boolean return value of the expression. 
```
self.cells.all? { |cell| cell == "X" || cell == "O" }
```

File: lib/board.rb [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+  end
+
+  def turn_count
+    counter = 0
+      self.cells.each do |cell| if cell == "X" || cell == "O"
+        counter += 1
+      end
+    end
+    counter
```
Comment (randallreedjr) [inline lib/board.rb]: Use the highest level and most expressive iterator possible to solve a problem. [`count`](http://ruby-doc.org/core-2.2.0/Array.html#method-i-count) on the cells is perfect and the rest of this method is not needed. Entire body could read:

```ruby
def turn_count
  cells.count {|cell| cell == "O" || cell == "X"}
end
```

File: lib/game.rb [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+                        [6,4,2]]
+  def initialize(player_1=nil, player_2=nil, board=nil)
+    @board = board || Board.new()
+    @player_1 = player_1 || Human.new("X")
+    @player_2 = player_2 || Human.new("O")
+    @winner = nil
+  end
+
+  def current_player
+    return @board.turn_count%2 == 0 ? @player_1 : @player_2
```
Comment (randallreedjr) [inline lib/game.rb]: Use [`Integer#even?`](http://ruby-doc.org/core-1.8.7/Integer.html#method-i-even-3F) or [`Integer#odd?`](http://ruby-doc.org/core-1.8.7/Integer.html#method-i-odd-3F)

`@board.turn_count.even? ? @player_1 : @player_2`

File: lib/game.rb [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+  end
+
+  def over?
+    if won?
+      return true
+    elsif draw?
+      return true
```
Comment (randallreedjr) [inline lib/game.rb]: This kinda looks like `won? || draw?` to me...

File: lib/game.rb [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+    else
+      return false
+    end
+  end
+
+  def won?
+    WIN_COMBINATIONS.each do |triplet|
+      if triplet.all?{|s| @board.cells[s]=="O"}
```
Comment (randallreedjr) [inline lib/game.rb]: Nice combination of iterators here to keep the code succinct.

File: lib/game.rb [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+        puts "Computer's move: "
+      end
+    else
+      turn
+    end
+  end
+
+  def play
+    #don't print out a board if the computer plays first vs human
+    if @board.turn_count==0 && current_player.is_a?(Computer)
```
Comment (randallreedjr) [inline lib/game.rb]: Good handling of an edge case.

File: lib/players/computer.rb [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+        #this will return a 1-element array
+        hotspot = triplet.select{|s| board.cells[s]==" "}
+        return hotspot.first + 1
+      end
+    end
+    return normal
+  end
+
+  def impossible
+    return rand(1..9).to_s
```
Comment (randallreedjr) [inline lib/players/computer.rb]: Perhaps not so impossible after all...

File: lib/players/human.rb [hunk 12]
```diff
@@ -0,0 +1,7 @@
+class Human < Player
+
+  def move(token, board=nil)
+    input = gets.strip
+    return input
```
Comment (randallreedjr) [inline lib/players/human.rb]: You don't need to store your return value in a local variable; just utilize the implicit return value of the expression.
```
def move(token, board=nil)
  gets.strip
end
```

**Comment by randallreedjr** [comment]
@bhollan Great job! I left a few comments, feel free to respond with any questions or comments. Closing this PR for now.

File: bin/tictactoe (hunk 2 above)
Comment (bhollan) [inline bin/tictactoe]: This was all @the-widget .  I loved it too.

Several times during testing: "right, yes, thank you. >:P"

File: lib/players/computer.rb (hunk 11 above)
Comment (bhollan) [inline lib/players/computer.rb]: We were both planning to do the minimax algorithm, but we also were/are both highly interested in our speed through the curriculum.  It was a strategic team choice.  :)

I don't know about @the-widget , but I plan to do minimax on this later and do a blog post about it.  Maybe even write a 'connect 4' version too.

File: lib/players/human.rb (hunk 12 above)
Comment (bhollan) [inline lib/players/human.rb]: I'm more used to it now, but this was my early 'ruby jitters'.  I'm more bent towards the python "explicit is better than implicit" way of specifying things.

File: bin/tictactoe (hunk 1 above)
Comment (bhollan) [inline bin/tictactoe]: Is that just 'common/best practices' or does that actually affect performance/security type stuff?

File: lib/board.rb (hunk 5 above)
Comment (bhollan) [inline lib/board.rb]: Duh.  Whoops.

File: lib/game.rb (hunk 7 above)
Comment (bhollan) [inline lib/game.rb]: That's my engineering background smudging my new skills.  :/

File: lib/game.rb (hunk 8 above)
Comment (bhollan) [inline lib/game.rb]: This was a copy/paste from the procedural TTT.  That one was a testing nightmare for me and this was how that ended up.  I like the look of `won? || draw?` for sure though.

File: lib/game.rb (hunk 10 above)
Comment (bhollan) [inline lib/game.rb]: This one took me half an hour to find, and about 30 seconds to fix.

File: lib/players/computer.rb (hunk 11 above)
Comment (randallreedjr) [inline lib/players/computer.rb]: @bhollan definitely understandable. I would be very interested in reading that blog post. Minimax still confuses me...

File: bin/tictactoe (hunk 1 above)
Comment (randallreedjr) [inline bin/tictactoe]: I don't think there are performance or security impacts, just separation of concerns/encapsulation from an Object Oriented design perspective.

**Pull request closed without merging**